### ISSUE TITLE
recent PR broke raster layers

### ISSUE THREAD

**Issue by z0d14c** [alex3165/react-mapbox-gl#156]
Hi!

This PR broke raster type layers:
https://github.com/alex3165/react-mapbox-gl/commit/07e95f73f6c3e18c31dfc2e5f8431835be53b415

The reason being that raster type layers' sources **do not have the setData function** . This means on Layer.render() that the app will break because of a `TypeError`.

the `Layer` is probably overly biased towards geojson sources in particular. line 220 of layer.ts (the problem line) assumes source is GeoJSONSource and makeFeature basically assumes that the layer has geometry/coordinates (Raster layers do not). This is made all the more confusing by the fact that we have a separate component specifically named GeoJSONLayer.

Will try and get a reasonable fix in ASAP, am open to discussion on how we can make the layer.ts more safe for both raster and non-raster type layers. Tagging @jluxenberg as he made the offending PR so I don't wanna mess his stuff up too much when I make my own PR. 😛

**Comment by jluxenberg** [comment]
Ah, I see! Sorry for breaking this.

I guess we could do something silly like add a case in render specifically for the 'raster' layer type? I don't know the codebase well enough to suggest something better.

- Jared

**Comment by alex3165** [comment]
@z0d14c Can we add a quick fix by doing for now ?
```
    const source = map.getSource(this.props.sourceId || this.id) as MapboxGL.GeoJSONSource;
    source.setData && source.setData({
      type: 'FeatureCollection',
      features
    });
```

**Comment by z0d14c** [comment]
@alex3165 that will most likely work as a quick fix. i will test locally and let you know.

@jluxenberg yeah, something like that is probably fine. ill take a closer look and update the thread if something more extensive is needed.

**Comment by alex3165** [comment]
I published a patch `1.4.1` with the solution I suggested, @z0d14c Please open a new issue if there is still a bug or feel free to open a PR if you find a better solution

**Comment by z0d14c** [comment]
Should be good for now, I'll look into getting a more robust pattern into
layer eventually
